checker/src/main/java/org/checkerframework/checker/index/upperbound/UpperBoundVisitor.java

```diff
@@ -308,6 +400,78 @@ private boolean relaxedCommonAssignmentCheck(
+     * expQual includes LTL(rhsSeq, from) for some rhsSeq that has HSS, and varLtlQual includes
+     * LTL(a), then the LTL(a) will be removed from varLtlQual
+     */
+    private UBQualifier processSubsequenceForRHS(LessThanLengthOf varLtlQual, UBQualifier expQual) {
+        UBQualifier newLHS = varLtlQual;
+        if (!expQual.isLessThanLengthQualifier()) {
+            return newLHS;
+        }
+        LessThanLengthOf rhsQual = (LessThanLengthOf) expQual;
+        for (String rhsSeq : rhsQual.getSequences()) {
+            FlowExpressions.Receiver rec =
+                    getReceiverFromJavaExpressionString(rhsSeq, atypeFactory, getCurrentPath());
+            Subsequence subSeq = Subsequence.getSubsequenceFromReceiver(rec, atypeFactory);
+            FlowExpressionParseUtil.FlowExpressionContext context =
+                    getContextFromReceiver(rec, checker);
+
+            if (subSeq != null) {
+                // viewpoint adapt strings from HasSubsequence expression
+                String from = standardizeAndViewpointAdapt(subSeq.from, getCurrentPath(), context);
+                String a = standardizeAndViewpointAdapt(subSeq.array, getCurrentPath(), context);
+
+                if (rhsQual.hasSequenceWithOffset(rhsSeq, from)) {
+                    if (varLtlQual.hasSequenceWithOffset(a, "0")) {
+                        newLHS = ((LessThanLengthOf) newLHS).removeOffset(a, 0);
+                    }
+                }
+            }
+        }
+        return newLHS;
+    }
+
+    /**
+     * If varLtlQual is @LTLength(lhsSeq) and "a" is a subsequence of lhsSeq starting at "from",
+     * then remove lhsSeq from a copy of varLtlQual and return the copy. Otherwise return
+     * varLtlQual.
+     */
+    private UBQualifier processSubsequenceForLHS(LessThanLengthOf varLtlQual, UBQualifier expQual) {
+        UBQualifier newLHS = varLtlQual;
+        for (String lhsSeq : varLtlQual.getSequences()) {
+            // check is lhsSeq is an actual LTL
+            if (!varLtlQual.hasSequenceWithOffset(lhsSeq, 0)) {
+                continue;
+            }
+            Receiver rec =
+                    getReceiverFromJavaExpressionString(lhsSeq, atypeFactory, getCurrentPath());
+            Subsequence subSeq = Subsequence.getSubsequenceFromReceiver(rec, atypeFactory);
+            if (subSeq != null) {
+                // viewpoint adapt strings from HasSubsequence expression
+                FlowExpressionContext context = getContextFromReceiver(rec, checker);
+                String from = standardizeAndViewpointAdapt(subSeq.from, getCurrentPath(), context);
+                String a = standardizeAndViewpointAdapt(subSeq.array, getCurrentPath(), context);
+
+                if (expQual.hasSequenceWithOffset(a, negateString(from))) {
+                    // this cast is safe because LTLs cannot contain duplicates
+                    newLHS = ((LessThanLengthOf) newLHS).removeOffset(lhsSeq, 0);
```

Comment:
If varLtlQual has more than one lhsSeq with a subsequence, then is only updates the qualifier for one of them.  (Previous updates to newLHS are overwritten.) It's probably fine to assume that only one lhsSeq has a subsequence, but it should be clear.  Also, you could just return the newLHS here.